### PR TITLE
server: shift the speculative batch indices on the sub-batch retry

### PR DESCRIPTION
**Pull request by danielhanchen**
`--spec-type draft-mtp` with `--parallel N --kv-unified` throws
`speculative batch index %d is not inside the current sub-batch` and kills the request,
whenever the KV cache is full enough that llama-server has to fall back to a smaller batch.

## The bug

When a decode fails to fit, the server retries the same batch in smaller pieces. On that
path it already does two things for `slot.i_batch`: skips a slot whose index lies outside
the current view, and shifts the index by the view's offset, with the comment "shifted
according to the current sub-batch". It does neither for `slot.spec_i_batch`.

That asymmetry is the whole defect. The speculative indices stay absolute while the view
they are read against becomes relative, so the bounds check fires on indices that were
correct when they were recorded.

The check is also the only thing standing between this and a wrong answer: with it removed
the same indices would sample from another slot's logits.

## What this changes

Three commits, smallest first:

1. Keep a speculative block inside one sub-batch, and shift its indices. A slot's
   `spec_i_batch` entries are contiguous, so a block that would straddle the end of a view
   cuts the view short instead, and the surviving indices are shifted by the offset exactly
   as `i_batch` is.
2. Never halve `n_batch` below one speculative block. The retry ladder halves the batch
   each time; without a floor it can end up smaller than a single block, which no amount of
   cutting can fit.
3. Do not let that floor trap the ladder. A block starting at the view's own offset and
   reaching past its end cannot be cut, so the floor must not stop the ladder from
   descending, or the retry never terminates.

The error message that remains is deliberately different: a block that genuinely cannot be
placed now says `speculative block [%d, %d] straddles the current sub-batch [%d, %d)`, so
the two situations are distinguishable in a log.

## Measurement

Two independent runs, because they answer different questions.

**The branch, against its own base.** Both arms built from this tree, differing only by these
three commits, run as a bare llama-server with no caller in front of it. Four concurrent
chats, `-c 16384 --parallel 4 --kv-unified --spec-type draft-mtp --spec-draft-n-max 2`,
3000-token prompts, driven until the pool overflows:

| arm | `is not inside the current sub-batch` | `straddles the current sub-batch` | chats lost |
|---|---|---|---|
| master | 8 | 0 | 4 of 4 |
| this branch | **0** | 4 | 4 of 4 |

**The same change on the base Unsloth Studio actually ships** (b10715-mix, 92cedc867),
interleaved control and treatment over two rounds with the order reversed, each arm confirmed
from `/proc/PID/maps` of the live server: 18 occurrences of the original error on control,
**0** on treatment.

## What this does and does not fix

It removes the bounds violation this issue is about. It is NOT a complete fix, and the table
above says so: two of the four chats still fail on this branch, with a different error.

The remaining case is a speculative block that begins exactly at the view's own offset and
reaches past its end, `[4, 5]` against `[4, 5)`. It cannot be cut, because cutting the view
before it leaves an empty view, and the retry ladder has already driven `n_batch` below one
block by the time it appears. The code reports it rather than sampling from logits that were
never computed, which is the safe half of the choice.

Sampling only the part of the block inside the view would look like an obvious fix and is
not one: the drafted tokens beyond the view are already in the batch at positions this
sub-batch does not decode, so discarding them mid-flight risks leaving stale cells, and the
failure mode of getting that wrong is a wrong answer rather than an error. I have not
convinced myself of a correct version, so I have not written one.

What the change buys, then, is narrower than "the bug is gone" and still worth having:

  * the failures that remain are a distinct, self-describing message, so the two situations
    can be told apart in a log rather than presenting as one bounds error;
  * on Studio's own base, under its own load, the count went 18 to 0, because the cases it
    handles are the common ones;
  * for a caller that keeps the cache below the line, as Unsloth Studio's preemption now
    does, none of these are reached at all. The value there is that the residual overshoot
    degrades to `Context size has been exceeded`, which a caller can see coming.

The natural next step is to make the unfittable case terminate that one slot through the
ordinary context-exceeded path instead of throwing, which composes with
unslothai/llama.cpp#183. That is a separate change and is not in this PR.

## Notes

`server-context.cpp` only. Builds clean, and the measurements above are from that build. No
effect without `--spec-type`: every new branch is inside a block guarded on a non-empty
`spec_i_batch`.

There is no automated CI on this PR, and that is expected rather than a gap: the only
`pull_request`-triggered workflow in this fork is the `pr-set.json` lint, which is gated on
paths this change does not touch. Prebuild checks arrive when a PR is pinned into
`scripts/unsloth/pr-set.json`, which also ships it into the nightly builds, so I have left
that decision to a maintainer.

### CODE REVIEW

**Comment by danielhanchen** [comment]
Added `945cdfc9`, which removes the last way this path can still end every conversation on the server.

### What was left

The commit before it stops the retry ladder from cutting through a speculative block, and reports the case it cannot fix by throwing. That throw reaches `abort_all_slots`, so one slot drafting into a full cache still ends every conversation. Measured on this branch against its own base: control 8 occurrences of the index error and 0 of 4 chats completing, treatment 0 index errors and 4 of its own throws, still 0 of 4 completing.

### What it does now

`spec_i_batch[0]` is the index of the token that was actually sampled last round rather than a draft, and in the straddling case it is inside the current view. So the slot gives up the prediction and samples its next token the ordinary way. The cost is that one step's speedup. It is only reachable once the ladder has driven `n_batch` below one block, which needs a cache full enough that the view is narrower than the draft.

### Measurement

Four concurrent chats, 3000-token prompts, `-c 16384 --parallel 4 --kv-unified --spec-type draft-mtp --spec-draft-n-max 2`, on this fork's own base:

| build | index errors | own throws | drafts dropped | chats completing |
|---|---|---|---|---|
| master | 8 | - | - | 0 of 4 |
| this branch before `945cdfc9` | 0 | 4 | - | 0 of 4 |
| this branch | 0 | 0 | - | 0 of 4 |
| this branch + #183 | 0 | 0 | 2 | 1 of 4 |

The solid result is the third row and the `drafts dropped` column: the fallback fires and nothing aborts. The `1 of 4` is suggestive rather than measured, since this harness varies widely run to run (single-chat token counts range from 1540 to 13924 across otherwise identical runs).

What remains in every row is genuine cache exhaustion. Four chats each holding a 3000-token prompt plus a long answer do not fit in 16384 cells, and a bare llama-server has no admission control to say so, which is what #183 narrows from "every slot" to "one slot". With a scheduler in front that reserves the drafts, the same four chats complete 4 of 4 with zero context-exceeded errors.
